includes/REST_API/Products_Controller.php

```diff
@@ -0,0 +1,489 @@
+ * @phpstan-type ShopifyGraphQLProductImage array{url: string, altText: string}
+ * @phpstan-type ShopifyGraphQLProduct array{id: string, title: string, vendor: string, description: string, onlineStoreUrl: string, images: array{edges: array{node: ShopifyGraphQLProductImage}[]}, priceRange: ShopifyGraphQLPriceRange}
+ * @phpstan-type ShopifyGraphQLResponse array{errors?: ShopifyGraphQLError, data: array{products: array{edges: array{node: ShopifyGraphQLProduct}[]}}}
+ */
+class Products_Controller extends REST_Controller implements HasRequirements {
+
+	/**
+	 * Settings instance.
+	 *
+	 * @var Settings Settings instance.
+	 */
+	private $settings;
+
+	/**
+	 * Story_Post_Type instance.
+	 *
+	 * @var Story_Post_Type Story_Post_Type instance.
+	 */
+	private $story_post_type;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param Settings        $settings Settings instance.
+	 * @param Story_Post_Type $story_post_type Story_Post_Type instance.
+	 */
+	public function __construct( Settings $settings, Story_Post_Type $story_post_type ) {
+		$this->settings        = $settings;
+		$this->story_post_type = $story_post_type;
+
+		$this->namespace = 'web-stories/v1';
+		$this->rest_base = 'products';
+	}
+
+	/**
+	 * Get the list of service IDs required for this service to be registered.
+	 *
+	 * Needed because the story post type needs to be registered first.
+	 *
+	 * @since 1.20.0
+	 *
+	 * @return string[] List of required services.
+	 */
+	public static function get_requirements(): array {
+		return [ 'settings', 'story_post_type' ];
+	}
+
+	/**
+	 * Registers routes for links.
+	 *
+	 * @since 1.20.0
+	 *
+	 * @see register_rest_route()
+	 */
+	public function register_routes(): void {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'get_items' ],
+					'permission_callback' => [ $this, 'permissions_check' ],
+					'args'                => $this->get_collection_params(),
+				],
+				'schema' => [ $this, 'get_public_item_schema' ],
+			]
+		);
+	}
+
+	/**
+	 * Checks if a given request has access to get and create items.
+	 *
+	 * @since 1.20.0
+	 *
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function permissions_check() {
+		if ( ! $this->story_post_type->has_cap( 'edit_posts' ) ) {
+			return new \WP_Error(
+				'rest_forbidden',
+				__( 'Sorry, you are not allowed to manage publisher logos.', 'web-stories' ),
+				[ 'status' => rest_authorization_required_code() ]
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Retrieves all products.
+	 *
+	 * @since 1.20.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
```

Comment:
Currently this endpoint does not respect fields that are passed to it. A `prepare_item_for_response` method should be added. Also `get_items_shopify` and `get_items_woocommerce` should really just return an array / object of data or WP_Error. Formatting of that data should happen outside of these methods, so this code is more testable and DRY.

Reply:
Yes, but that's something for #11154, as indicated in the PR description. This code is still very rough because of that.